### ISSUE TITLE
How to set an OSI for a given DS ?

### ISSUE THREAD

**Issue by vacary** [siconos/siconos#142]
Following the comment of @radarsat1  on my commit f57b5ac9caf9fe0f5494f376fcb45841c30a5edc 
I think it could be useful to discuss how to set the OSI for a given DS

**Comment by radarsat1** [comment]
Yes, I started adding calls to Simulation::setNonSmoothDynamicalSystemPtr() after the Simulation was created, because I figured that it didn't really make sense to have a Simulation without an NSDS anyways.  However, maybe it's a non-issue since prepareIntegratorForDS actually takes the Model as a parameter anyways, so we could just take the NSDS from there.  I didn't do it that way since the idea was that supplying the Model should be optional.  In principle we should also assert() that the Simultion::_nsds == Model::_nsnds.

In any case, the steps needed for adding a DS to the simulation and model are, as I understand them,

1. Add the DS to the graph using NSDS::insertDynamicalSystem()
2. Ensure the OSI is part of Simulation::_allOSI using Simulation::insertIntegrator()
3. Set the OSI for the DS via topology()->setOSI()
4. initialize DS work vectors using OSI::initializeDynamicalSystem(), which requires the model and time.

The ambiguity in the API comes from the fact that the NSDS is shared between Simulation and Model.  I put prepareIntegratorForDS() into the Simulation since it requires to access _allOSI, but since it also needs the NSDS which it not guaranteed for Simulation, we have to choose between the following options:

1. Take the NSDS from the Model, require the Model is provided.
2. Ensure Simulation has an NSDS
3. Move prepareIntegratorForDS to Model, provide Simulation as argument.

I have been doing (2), but we could do (1) instead.  (3) does not make sense to me.

**Comment by vacary** [comment]
I am ok with the 4 steps you described for adding a DS into the simulation.

For the API, perhaps we can revert one step by implementing a method such as

~~~
OneStepIntegrator::addDynamicalSystemToIntegrate(DynamicalSystem & ds)
{
topology()->setOSI( *this, ds);
}
~~~

If the first two steps has been done successfully, then the topology is reachable but the OSI.

@radarsat1 What do you think ?

**Comment by radarsat1** [comment]
Hm, I didn't realize the OSI also has a pointer to the NSDS.  Could be interesting, I will think a bit more.  My first impression is that it doesn't sound very intuitive to "add" the DS to the OSI, but in fact if the OSI needs workspace for each DS then maybe that makes sense.

**Comment by vacary** [comment]
add is perhaps not the correct word.

better idea ? 
OneStepIntegrator::IntegrateDynamicalSystem(DynamicalSystem & ds)
OneStepIntegrator::dealWithDynamicalSystem(DynamicalSystem & ds)
OneStepIntegrator::processDynamicalSystem(DynamicalSystem & ds)

**Comment by vacary** [comment]
Hm, the problem will be the same with my pseudo solution
OneStepIntegrator::addDynamicalSystemToIntegrate(DynamicalSystem & ds)
We will need to have the nsds in the OSI before calling the method. So we will need to have to call model initialization before.

**Comment by radarsat1** [comment]
Ok I agree that is the best option.  It could even be possible to require passing the nonSmoothDynamicalSystem() to prepareIntegratorForDS(), which makes even more sense, because then you are directly saying where to store the link.  But we need the model anyway for the call to OSI->initializeDynamicalSystem().  (Again though, the Model isn't actually used in any implementation..)

I am trying it now and running tests, I will make a PR.  If you want still to keep setOSI for 4.1 since we are still understanding the API change that is reasonable (easy to change back), but I think this is sort of a sensible compromise: two calls for adding a DS, one for the Model, one for the Simulation.

**Comment by radarsat1** [comment]
> easy to change back

just to note, this would mean just calling setOSI and making sure that OSI::initializeDynamicalSystem() is called when needed. i.e. if added after Simulation::initialize().

**Comment by radarsat1** [comment]
I am fixing up many examples to work with this and I find it really disturbs the separation of Model and Simulation.  In order to assign an OSI to a DS, one needs to already have the Simulation created, which  in many cases is only done after the Model is created and the graph fully populated.  So it strikes me that in Siconos, the assignment of a DS to an Integrator is considered part of the Model, not part of the Simulation.  (The OSI is the only part of the Simulation typically created before the Model.)

In fact there is now nothing in prepareIntegratorForDS that requires it to be in Simulation, except for adding the OSI to `_allOSI`.

What if prepareIntegratorForDS() was simply part of Model?  We could even call it setOSI().

My real preference would be to remove the unused Model parameter from initializeDynamicalSystem(), so that it can be called from NSDS::insertDynamicalSystem() by adding an OSI parameter.  Perhaps I will make an alternative PR for that, and we can compare.

**Comment by xhub** [comment]
My limited understanding is that there are some concepts and design choices behind those names. I'm not able to locate a good reference for that.

My impression was that people agree that a better separation would be a good addition. A clear roadmap would be great

**Comment by radarsat1** [comment]
I agree that the separation is very important.  The one exception seems to be that the links between DS and OSI are typically created in the Model, before the Simulation is created.  After fixing up some examples, I see that prepareIntegratorForDS being a member of Simulation makes this worse, as now the Simulation needs to exist while populating the Model.

I also discovered that I was wrong about being able to remove Model from initializeDynamicalSystem, it is indeed used in ZeroOrderHoldOSI to create the MatrixIntegrator object.  This needs the Model times (t0 and tFinal), and the Simulation pointer, to copy the TimeDiscretisation.  So perhaps calling initializeDynamicalSystem from NSDS::setOSI without passing in the Model is not possible.  Too bad, I was hoping it would be a cleaner solution.  But I see that it makes sense for the case that we create sub-integrators for each object.  Are we completely against adding setOSI() to Model?  That seems to be the one place where everything needed to initialize a DS/OSI link is available.

Maybe @vacary is right to leave this for 4.2, I can prepare a PR calling setOSI and initializeDynamicalSystem explicitly wherever needed so that we can get on with 4.1.

**Comment by xhub** [comment]
I don't think that ZeroOrderHoldOSI should be a blocker for such a change. I coded that OSI and I'm pretty sure we can find a workaround. There is surely a better way to pass the information needed for the MatrixIntegrator object.

**Comment by radarsat1** [comment]
imho it would be a bit "cleaner" to have no need to reference the Model when assigning an Integrator/DS link and initializing memory for it, which only needs to modify the graph properties, so yeah maybe..  it does seem to be the only place as far as I can tell.  Care to take a shot at it?  I can try again next week.  I know it's a bit silly to spend time on but I do think it will be useful to get this "right" in the long run.  For instance I guess one long-term goal is to be able to run multiple Simulations from the same Model + initial conditions, if I understand correctly.

Perhaps we could add some lazier initialization to ZeroOrderHoldOSI.

**Comment by radarsat1** [comment]
I will add that when I met with the OSRF (the Gazebo people), one engineer specifically asked me whether Siconos had a feature to be able to snapshot / rewind time.  I said no, but I do think there is potential for it..

**Comment by vacary** [comment]
For the moment, I prefer the strategy implement in pull request #166.

Since the location of the Topology is a little ambiguous since it is in the middle of modeling and simulation, it causes a lot of problems for  setting some properties such as psi before the initialization. We have to think more thoroughly to this point. 

I suggest to merge PR #166 and to  keep this issue open but with milestone 4.2

We discussed a long a time ago to have a Topology in the modeling side to register the ds and the interaction and their link.  A "simulation" version could also be useful but I don't know how to implement it.

For the setOSI issue, the simplest way is to store temporary this de concerned by the OSI into a private map with  no accessor. Then we can implement something like:

~~~

SP::NewtonEulerDS ds1(new NewtonEulerDS(q10, v10, m));
SP::MoreauJeanOSI OSI(new MoreauJeanOSI(theta));
OSI->handle(ds1) // store into a temp map into OSI

~~~

 When we initialize the OSI we must have the simulation ans the topology at hand, we than transfer the info

**Comment by radarsat1** [comment]
That makes some sense. I will merge #166.

**Comment by vacary** [comment]
Let us try
```
SP::NewtonEulerDS ds1(new NewtonEulerDS(q10, v10, m));
SP::MoreauJeanOSI OSI(new MoreauJeanOSI(theta));
OSI->associate(ds1) // store into a temp map into OSI
```
